src/matches.rs

```diff
@@ -71,101 +96,204 @@ impl LateLintPass for MatchPass {
+                                expr_block(cx, false_expr, None, "..")));
                     } else {
-                        span_lint(cx, MATCH_BOOL, expr.span,
+                        span_help_and_lint(cx, MATCH_BOOL, expr.span,
                             "you seem to be trying to match on a boolean expression. \
-                            Consider using an if..else block");
+                           Consider using an if..else block:",
+                           &format!("try\nif {} {}",
+                                snippet(cx, ex.span, "b"),
+                                expr_block(cx, true_expr, None, "..")));
                     }
+                } else if !is_unit_expr(false_expr) {
+                    span_help_and_lint(cx, MATCH_BOOL, expr.span,
+                        "you seem to be trying to match on a boolean expression. \
+                       Consider using an if..else block:",
+                       &format!("try\nif !{} {}",
+                            snippet(cx, ex.span, "b"),
+                            expr_block(cx, false_expr, None, "..")));
                 } else {
                     span_lint(cx, MATCH_BOOL, expr.span,
-                        "you seem to be trying to match on a boolean expression. \
-                        Consider using an if..else block");
+                           "you seem to be trying to match on a boolean expression. \
+                           Consider using an if..else block");
                 }
+            } else {
+                span_lint(cx, MATCH_BOOL, expr.span,
+                    "you seem to be trying to match on a boolean expression. \
+                    Consider using an if..else block");
             }
+        } else {
+            span_lint(cx, MATCH_BOOL, expr.span,
+                "you seem to be trying to match on a boolean expression. \
+                Consider using an if..else block");
         }
-        if let ExprMatch(ref ex, ref arms, source) = expr.node {
-            // check preconditions for MATCH_REF_PATS
-            if has_only_ref_pats(arms) {
-                if let ExprAddrOf(Mutability::MutImmutable, ref inner) = ex.node {
-                    let template = match_template(cx, expr.span, source, "", inner);
-                    span_lint(cx, MATCH_REF_PATS, expr.span, &format!(
-                        "you don't need to add `&` to both the expression \
-                         and the patterns: use `{}`", template));
-                } else {
-                    let template = match_template(cx, expr.span, source, "*", ex);
-                    span_lint(cx, MATCH_REF_PATS, expr.span, &format!(
-                        "instead of prefixing all patterns with `&`, you can dereference the \
-                         expression: `{}`", template));
-                }
+    }
+}
+
+fn check_overlapping_arms(cx: &LateContext, arms: &[Arm]) {
+    if arms.len() >= 2 {
```

Comment:
We should only run this lint if the type of the match is an integral one. Type lookups are cheap, but collecting all the arms into a vector and all are not, so it's better to exit early and avoid all the churn if the match isn't of the correct type.